// @gtm-support/vue2-gtm/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useGtm = exports.GtmPlugin = exports.loadScript = exports.hasScript = exports.GtmSupport = exports.assertIsGtmId = void 0;
var core_1 = require("@gtm-support/core");
Object.defineProperty(exports, "GtmPlugin", { enumerable: true, get: function () { return core_1.GtmSupport; } });
var gtmPlugin;
/**
 * Installation procedure.
 *
 * @param Vue The Vue instance.
 * @param options Configuration options.
 */
function install(Vue, options) {
    if (options === void 0) { options = { id: '' }; }
    // Apply default configuration
    options = __assign({ trackOnNextTick: false }, options);
    // Add to vue prototype and also from globals
    gtmPlugin = new core_1.GtmSupport(options);
    Vue.prototype.$gtm = Vue.gtm = gtmPlugin;
    // Handle vue-router if defined
    if (options.vueRouter) {
        initVueRouterGuard(Vue, options.vueRouter, options.ignoredViews, options.trackOnNextTick);
    }
    // Load GTM script when enabled
    if (gtmPlugin.options.enabled && gtmPlugin.options.loadScript) {
        if (Array.isArray(options.id)) {
            options.id.forEach(function (id) {
                if (typeof id === 'string') {
                    core_1.loadScript(id, options);
                }
                else {
                    var newConf = __assign({}, options);
                    if (id.queryParams != null) {
                        newConf.queryParams = __assign(__assign({}, newConf.queryParams), id.queryParams);
                    }
                    core_1.loadScript(id.id, newConf);
                }
            });
        }
        else {
            core_1.loadScript(options.id, options);
        }
    }
}
/**
 * Initialize the router guard.
 *
 * @param Vue The Vue instance.
 * @param vueRouter The Vue router instance to attach the guard.
 * @param ignoredViews An array of route name that will be ignored.
 * @param trackOnNextTick Whether or not to call `trackView` in `Vue.nextTick`.
 */
function initVueRouterGuard(Vue, vueRouter, ignoredViews, trackOnNextTick) {
    if (ignoredViews === void 0) { ignoredViews = []; }
    if (!vueRouter) {
        console.warn("[VueGtm]: You tried to register 'vueRouter' for vue-gtm, but 'vue-router' was not found.");
        return;
    }
    vueRouter.afterEach(function (to, from) {
        var _a, _b, _c;
        // Ignore some routes
        if (typeof to.name !== 'string' ||
            (Array.isArray(ignoredViews) && ignoredViews.includes(to.name)) ||
            (typeof ignoredViews === 'function' && ignoredViews(to, from))) {
            return;
        }
        // Dispatch vue event using meta gtm value if defined otherwise fallback to route name
        var name = to.meta && typeof to.meta.gtm === 'string' && !!to.meta.gtm ? to.meta.gtm : to.name;
        var additionalEventData = (_b = (_a = to.meta) === null || _a === void 0 ? void 0 : _a.gtmAdditionalEventData) !== null && _b !== void 0 ? _b : {};
        var baseUrl = (_c = vueRouter.options.base) !== null && _c !== void 0 ? _c : '';
        var fullUrl = baseUrl;
        if (!fullUrl.endsWith('/')) {
            fullUrl += '/';
        }
        fullUrl += to.fullPath.startsWith('/') ? to.fullPath.substr(1) : to.fullPath;
        if (trackOnNextTick) {
            Vue.nextTick(function () {
                gtmPlugin === null || gtmPlugin === void 0 ? void 0 : gtmPlugin.trackView(name, fullUrl, additionalEventData);
            });
        }
        else {
            gtmPlugin === null || gtmPlugin === void 0 ? void 0 : gtmPlugin.trackView(name, fullUrl, additionalEventData);
        }
    });
}
var _default = { install: install };
var core_2 = require("@gtm-support/core");
Object.defineProperty(exports, "assertIsGtmId", { enumerable: true, get: function () { return core_2.assertIsGtmId; } });
Object.defineProperty(exports, "GtmSupport", { enumerable: true, get: function () { return core_2.GtmSupport; } });
Object.defineProperty(exports, "hasScript", { enumerable: true, get: function () { return core_2.hasScript; } });
Object.defineProperty(exports, "loadScript", { enumerable: true, get: function () { return core_2.loadScript; } });
exports.default = _default;
/**
 * Returns GTM plugin instance to be used via Composition API inside setup method.
 *
 * @returns The Vue GTM instance if the it was installed, otherwise `undefined`.
 */
function useGtm() {
    return gtmPlugin;
}
exports.useGtm = useGtm;
